README


This archive contains the distribution Email-MIME,
version 1.928:

  easy MIME message handling

This software is copyright (c) 2004 by Simon Cozens and Casey West.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


This README file was generated by Dist::Zilla::Plugin::Readme v5.029.



Changes
Revision history for Perl extension Email::MIME.

1.928     2014-12-15 21:47:49-05:00 America/New_York
        - when converting from single part to multipart, drop
          Content-Transfer-Encoding and Content-Type's charset property
        - update subparts in walk_parts if stringification changes

1.927     2014-12-04 10:19:58-05:00 America/New_York
        - add header_str_pairs; like header_pairs, but decoded
        - remove Email:MIME::Creator from synopsis

1.926     2014-01-28 23:28:41-05:00 America/New_York
        - improve EOL processing with quoted-printable (thanks, Tomohiro
          Hosaka!)
        - avoid uninitialized warnings [rt.cpan.org #91707]

1.925     2013-11-08 07:01:03 America/New_York
        - improve reliability of mail reconstruction during walk_parts (thanks,
          Tomohiro Hosaka!)

1.924     2013-08-10 23:23:55 America/New_York
          update use of Email::MIME::ContentType to match new, fixed hash keys:
          type/subtype

1.923     2013-08-08 21:59:02 America/New_York
          do not consider the part-ending CRLF part of the body

          avoid undefined warnings in debug_structure [rt.cpan.org #82388]
          (Thanks, Kurt Anderson)

          better error message when the given body is a ref but not a scalar
          ref [rt.cpan.org #59205]

1.922     2013-07-10 08:45:02 America/New_York
          repackage, fixing version number

1.921     2013-07-01 22:51:01 America/New_York
          repackage, remove PEP links, update bugtracker

1.920     2013-06-17
          do not call parts_set during walk_parts unless the parts have
          actually changed

          When trying to decode a body, fall back to 7bit if the encoding is
          unknown.  Trying to create a new body in an unknown encoding is still
          forbidden.  This should make it easier to handle broken messages
          from "the internet."  ("Content-Transfer-Encoding: n"!?)

1.912_01  2013-04-08
          try to encode headers based on the header structure, if it has one,
          rather than treating the header as a big string in all cases; thanks
          for this work go to Jesse Luehrs

1.911     2012-07-22
          require a newer Email::Simple and cope with its improved line ending
          tweaks

1.910     2011-09-12

          document the header_str arg to ->create more thoroughly (i.e., at
          all)

1.909     2011-09-08

          dial the perl prereq back to 5.8.1 by popular demand... actually

1.908     2011-06-01

          dial the perl prereq back to 5.8.1 by popular demand

1.907     2011-02-02

          require 5.8.5 for sane encoding

          be more lenient with Content-Tranfser-Encoding values; stop at
          semicolons

1.906     2010-10-07
          update our prereq on Test::More to 0.88

1.905     2010-09-06
          fix a long-standing bug that prevent walk_parts from reaching all
          parts (thanks, Erik Logtenberg)

1.904     2010-09-04
          eliminate memory leak in walk_parts (reported by MKANAT, RT #59581)

1.903     2009-12-23
          correct typo in body_set_str (RT #53004) (thanks, Herbert Leitz)

1.902     2009-11-11
          allow for padding spaces in the Content-Transfer-Encoding header
          (Geraint Edwards)

1.901     2009-11-05
          bump up Email::MIME::Encodings version required

1.900     2009-11-03
          merge in Email-MIME-Modifier and Email-MIME-Creator
          add better support for Unicode with body_str, header_str_set, etc.

1.863     2009-01-30
          no code changes
          add strangely missing copyright information

1.862     2009-01-22
          add repository location metadata
          always require Encode, never MIME::Words; this means that using
            Email::MIME on pre-5.008 will be difficult, if not impossible
          if a header can't be decoded, fall back to the raw header
          move decoding methods to Email::MIME::Header, add header_raw

1.861     2007-11-05
          added perl-minver.t -- Email::MIME requires perl >= 5.006
          we now require Email::Simple 2.003

1.860     2007-07-13
          tentative tweak to tests and C-T-E handling for charset
          probably needs more research, testing, and fixing

1.859     2007-03-20
          minor test improvements
          add "subparts" method

1.858     2007-02-09
          fixed laxity which allowed leading newlines in parts passed to 'new'
          for instantiation; this only broke when used with Email::Simple 1.998

          Shout out to all my former co-workers from General Cinemas unit 858!
          This version is for you.

1.857     2006-11-27
          further severing of Email::MIME from ::Simple's guts

1.856     2006-11-27
          reduce the extent to which Email::MIME touches Email::Simple guts
          improve documentation a bit
          improve packaging

1.855     2006-11-19
1.854     2006-11-12
          fix as_string to work with latest Email::Simple

1.853     2006-11-12
          always call header in scalar context when getting content type
          (thanks, JESSE)

1.852     2006-09-05
          avoid some warnings with undef headers or bodies

1.851     2006-08-22
          test coverage and layout improvements by SSORICHE

1.85      2006-07-13
          avoid memory leak due to circular references
            (bug 7007, starter patch from MARCB)
          pod tests

1.81      2004-10-29
          Fix bug in headers, now returns list in list context.
          Dependency updates.
          Author change.

1.8       Wed Aug 25 10:47:15 BST 2004
          Split filename and invent_filename to separate methods, for the
          Stripper

1.6       Sun Jul  4 21:15:07 BST 2004
          Another fix by Casey - this time, don't decode on as_string

1.5       Thu Jul  1 18:38:06 BST 2004
          That didn't actually fix the bug at all.

1.4       Thu Jul  1 18:35:32 BST 2004
          Only parse attributes if there are some!


lib/Email/MIME/Encode.pm
use strict;
use warnings;
package Email::MIME::Encode;
# ABSTRACT: a private helper for MIME header encoding
$Email::MIME::Encode::VERSION = '1.928';
use Email::Address;
use Encode ();
use MIME::Base64();

my %encoders = (
    'Date'        => \&_date_time_encode,
    'From'        => \&_mailbox_list_encode,
    'Sender'      => \&_mailbox_encode,
    'Reply-To'    => \&_address_list_encode,
    'To'          => \&_address_list_encode,
    'Cc'          => \&_address_list_encode,
    'Bcc'         => \&_address_list_encode,
    'Message-ID'  => \&_msg_id_encode,
    'In-Reply-To' => \&_msg_id_encode,
    'References'  => \&_msg_id_encode,
    'Subject'     => \&_unstructured_encode,
    'Comments'    => \&_unstructured_encode,
);

sub maybe_mime_encode_header {
    my ($header, $val, $charset) = @_;

    return $val unless defined $val;
    return $val unless $val =~ /\P{ASCII}/
                    || $val =~ /=\?/;

    $header =~ s/^Resent-//;

    return $encoders{$header}->($val, $charset)
        if exists $encoders{$header};

    return _unstructured_encode($val, $charset);
}

sub _date_time_encode {
    my ($val, $charset) = @_;
    return $val;
}

sub _mailbox_encode {
    my ($val, $charset) = @_;
    return _mailbox_list_encode($val, $charset);
}

sub _mailbox_list_encode {
    my ($val, $charset) = @_;
    my @addrs = Email::Address->parse($val);

    @addrs = map {
        my $phrase = $_->phrase;
        $_->phrase(mime_encode($phrase, $charset))
            if defined $phrase && $phrase =~ /\P{ASCII}/;
        my $comment = $_->comment;
        $_->comment(mime_encode($comment, $charset))
            if defined $comment && $comment =~ /\P{ASCII}/;
        $_;
    } @addrs;

    return join(', ', map { $_->format } @addrs);
}

sub _address_encode {
    my ($val, $charset) = @_;
    return _address_list_encode($val, $charset);
}

sub _address_list_encode {
    my ($val, $charset) = @_;
    return _mailbox_list_encode($val, $charset); # XXX is this right?
}

sub _msg_id_encode {
    my ($val, $charset) = @_;
    return $val;
}

sub _unstructured_encode {
    my ($val, $charset) = @_;
    return mime_encode($val, $charset);
}

# XXX this is copied directly out of Courriel::Header
# eventually, this should be extracted out into something that could be shared
sub mime_encode {
    my $text    = shift;
    my $charset = Encode::find_encoding(shift)->mime_name();

    my $head = '=?' . $charset . '?B?';
    my $tail = '?=';

    my $base_length = 75 - ( length($head) + length($tail) );

    # This code is copied from Mail::Message::Field::Full in the Mail-Box
    # distro.
    my $real_length = int( $base_length / 4 ) * 3;

    my @result;
    my $chunk = q{};
    while ( length( my $chr = substr( $text, 0, 1, '' ) ) ) {
        my $chr = Encode::encode( $charset, $chr, 0 );

        if ( length($chunk) + length($chr) > $real_length ) {
            push @result, $head . MIME::Base64::encode_base64( $chunk, q{} ) . $tail;
            $chunk = q{};
        }

        $chunk .= $chr;
    }

    push @result, $head . MIME::Base64::encode_base64( $chunk, q{} ) . $tail
        if length $chunk;

    return join q{ }, @result;
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Email::MIME::Encode - a private helper for MIME header encoding

=head1 VERSION

version 1.928

=head1 AUTHORS

=over 4

=item *

Ricardo SIGNES <rjbs@cpan.org>

=item *

Casey West <casey@geeknest.com>

=item *

Simon Cozens <simon@cpan.org>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2004 by Simon Cozens and Casey West.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Email/MIME/Header.pm
use strict;
use warnings;
package Email::MIME::Header;
# ABSTRACT: the header of a MIME message
$Email::MIME::Header::VERSION = '1.928';
use parent 'Email::Simple::Header';

use Email::MIME::Encode;
use Encode 1.9801;

#pod =head1 DESCRIPTION
#pod
#pod This object behaves like a standard Email::Simple header, with the following
#pod changes:
#pod
#pod =for :list
#pod * the C<header> method automatically decodes encoded headers if possible
#pod * the C<header_raw> method returns the raw header; (read only for now)
#pod * stringification uses C<header_raw> rather than C<header>
#pod
#pod Note that C<header_set> does not do encoding for you, and expects an
#pod encoded header.  Thus, C<header_set> round-trips with C<header_raw>,
#pod not C<header>!  Be sure to properly encode your headers with
#pod C<Encode::encode('MIME-Header', $value)> before passing them to
#pod C<header_set>.
#pod
#pod Alternately, if you have Unicode (character) strings to set in headers, use the
#pod C<header_str_set> method.
#pod
#pod =cut

sub header {
  my $self  = shift;
  my $wanta = wantarray;

  return unless defined $wanta; # ??

  my @header = $wanta ? $self->SUPER::header(@_)
                      : scalar $self->SUPER::header(@_);

  local $@;
  foreach my $header (@header) {
    next unless defined $header;
    next unless $header =~ /=\?/;

    _maybe_decode(\$header);
  }
  return $wanta ? @header : $header[0];
}

sub header_raw {
  Carp::croak "header_raw may not be used to set headers" if @_ > 2;
  my ($self, $header) = @_;
  return $self->SUPER::header($header);
}

sub header_str_set {
  my ($self, $name, @vals) = @_;

  my @values = map {
    Email::MIME::Encode::maybe_mime_encode_header($name, $_, 'UTF-8')
  } @vals;

  $self->header_set($name => @values);
}

sub header_str_pairs {
  my ($self) = @_;

  my @pairs = $self->header_pairs;
  for (grep { $_ % 2 } (1 .. $#pairs)) {
    _maybe_decode(\$pairs[$_]);
  }

  return @pairs;
}

sub _maybe_decode {
  my ($str_ref) = @_;

  # The eval is to cope with unknown encodings, like Latin-62, or other
  # nonsense that gets put in there by spammers and weirdos
  # -- rjbs, 2014-12-04
  my $new;
  $$str_ref = $new
    if eval { $new = Encode::decode("MIME-Header", $$str_ref); 1 };
  return;
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Email::MIME::Header - the header of a MIME message

=head1 VERSION

version 1.928

=head1 DESCRIPTION

This object behaves like a standard Email::Simple header, with the following
changes:

=over 4

=item *

the C<header> method automatically decodes encoded headers if possible

=item *

the C<header_raw> method returns the raw header; (read only for now)

=item *

stringification uses C<header_raw> rather than C<header>

=back

Note that C<header_set> does not do encoding for you, and expects an
encoded header.  Thus, C<header_set> round-trips with C<header_raw>,
not C<header>!  Be sure to properly encode your headers with
C<Encode::encode('MIME-Header', $value)> before passing them to
C<header_set>.

Alternately, if you have Unicode (character) strings to set in headers, use the
C<header_str_set> method.

=head1 AUTHORS

=over 4

=item *

Ricardo SIGNES <rjbs@cpan.org>

=item *

Casey West <casey@geeknest.com>

=item *

Simon Cozens <simon@cpan.org>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2004 by Simon Cozens and Casey West.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Email/MIME/Creator.pm
use 5.008001;
use strict;
use warnings;
package Email::MIME::Creator;
# ABSTRACT: obsolete do-nothing library
$Email::MIME::Creator::VERSION = '1.928';
use parent q[Email::Simple::Creator];
use Email::MIME;
use Encode ();

sub _construct_part {
  my ($class, $body) = @_;

  my $is_binary = $body =~ /[\x00\x80-\xFF]/;

  my $content_type = $is_binary ? 'application/x-binary' : 'text/plain';

  Email::MIME->create(
    attributes => {
      content_type => $content_type,
      encoding     => ($is_binary ? 'base64' : ''),  # be safe
    },
    body => $body,
  );
}

1;

#pod =head1 SYNOPSIS
#pod
#pod You don't need to use this module for anything.
#pod
#pod =cut

__END__

=pod

=encoding UTF-8

=head1 NAME

Email::MIME::Creator - obsolete do-nothing library

=head1 VERSION

version 1.928

=head1 SYNOPSIS

You don't need to use this module for anything.

=head1 AUTHORS

=over 4

=item *

Ricardo SIGNES <rjbs@cpan.org>

=item *

Casey West <casey@geeknest.com>

=item *

Simon Cozens <simon@cpan.org>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2004 by Simon Cozens and Casey West.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.029.
use strict;
use warnings;

use 5.008001;

use ExtUtils::MakeMaker;



my %WriteMakefileArgs = (
  "ABSTRACT" => "easy MIME message handling",
  "AUTHOR" => "Ricardo SIGNES <rjbs\@cpan.org>, Casey West <casey\@geeknest.com>, Simon Cozens <simon\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Email-MIME",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.008001",
  "NAME" => "Email::MIME",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Email::Address" => 0,
    "Email::MIME::ContentType" => "1.016",
    "Email::MIME::Encodings" => "1.314",
    "Email::MessageID" => 0,
    "Email::Simple" => "2.102",
    "Email::Simple::Creator" => 0,
    "Email::Simple::Header" => 0,
    "Encode" => "1.9801",
    "MIME::Base64" => 0,
    "MIME::Types" => "1.13",
    "Scalar::Util" => 0,
    "parent" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Symbol" => 0,
    "Test::More" => "0.96",
    "utf8" => 0
  },
  "VERSION" => "1.928",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "Email::Address" => 0,
  "Email::MIME::ContentType" => "1.016",
  "Email::MIME::Encodings" => "1.314",
  "Email::MessageID" => 0,
  "Email::Simple" => "2.102",
  "Email::Simple::Creator" => 0,
  "Email::Simple::Header" => 0,
  "Encode" => "1.9801",
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "MIME::Base64" => 0,
  "MIME::Types" => "1.13",
  "Scalar::Util" => 0,
  "Symbol" => 0,
  "Test::More" => "0.96",
  "parent" => 0,
  "strict" => 0,
  "utf8" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



